llvm/lib/Transforms/Utils/InlineFunction.cpp

```diff
@@ -1674,8 +1674,9 @@ static Value *HandleByValArgument(Type *ByValType, Value *Arg,
   if (ByValAlignment)
     Alignment = std::max(Alignment, *ByValAlignment);
 
-  AllocaInst *NewAlloca = new AllocaInst(ByValType, DL.getAllocaAddrSpace(),
```

Comment:
Digging a bit into git log the patch that added DL.getAllocaAddrSpace() was the following:
```
commit 3c1fc768ed6cf2b01463df036ae6ae3b1f0de632
Author: Matt Arsenault <Matthew.Arsenault@amd.com>
Date:   Mon Apr 10 22:27:50 2017 +0000

    Allow DataLayout to specify addrspace for allocas.
```
and reasoning behind that patch:
"    These are problems for AMDGPU because alloca is used to
    implement the private address space, which uses a 32-bit
    index as the pointer value. Other pointers are 64-bit
    and behave more like LLVM's notion of generic address
    space. By changing the address space used for allocas,
    we can change our generic pointer type to be LLVM's generic
    pointer type which does have similar properties."
 
 makes sense to me especially in OpenCL context, where default address space is 4 (generic) and LLVM passes must respect target's datalayout when picking possible types for the created instructions.
 
Current fix brings some complications though, if byval pointer of a function was is default aka genetic address space, then the alloca would be in generic. Meanwhile per SPIR-V (to which LLVM IR is being lowered for OpenCL 2.0 and earlier) specification OpVariable must have non-generic storage class. It can probably be workarounded in https://github.com/KhronosGroup/SPIRV-LLVM-Translator by translating alloca addrspace(4) to OpVariable with Private storage class bringing extra address space casts or by modifying clang, introducing an exception for byval pointers making them not be in default address space, but in private, but I wonder if https://github.com/llvm/llvm-project/issues/97086 can be solved differently - instead here if Arg->getType()->getPointerAddressSpace() != DL.getAllocaAddrSpace() - we still create an alloca in default datalayout address space and add an AS cast, WDYT?

Reply:
> instead here if Arg->getType()->getPointerAddressSpace() != DL.getAllocaAddrSpace() - we still create an alloca in default datalayout address space and add an AS cast, WDYT?

I'm not sure I fully understand the proposal, but for the CHERI use case (the AMD version was based on the CHERI fork, which we aim to start upstreaming soon), allocas must be created in the alloca address space, not some other address space and cast.

Reply:
I mean: leave alloca in DL.getAllocaAddrSpace() AS and if it doesn't match with byval pointer's address space - create a cast and update users of the alloca to use the cast. But digging more into it might be not as trivial as I though initially - at least it seems it shouldn't be done in HandleByValArgument, but probably later in the stack.

Reply:
That would not break anything for CHERI, if it works for other use cases.

Reply:
No, this is the correct fix. We should follow what the existing IR says. The rule is not all allocas must have the alloca address space, it's the alloca address space is the default for contextless new values. 

As far as OpenCL is concerned such IR would have been malformed in the first place. If we wanted to handle this as a generic pointer with a private backed alloca, it should be handled by lowering the alloca with the wrong address space to include the cast (which is essentially what NVPTX does, it's just more work for the backend and less optimizable IR)